File: WorldFeed/Server/src/WorldFeed.Controllers/Tweet/TweetQueryGenerator.cs
﻿
namespace WorldFeed.Controllers.Tweet
{
    using System.Globalization;
    using System.Linq;
    using System.Text;

    using WorldFeed.Common.Extensions;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Parameters.TweetsClient;
    using WorldFeed.Common.Public.Parameters.TweetsClients;
    using WorldFeed.Common.QueryGenerators;
    using WorldFeed.Common.Settings;
    using WorldFeed.Controllers.Properties;
    using WorldFeed.Controllers.Shared;


    public class TweetQueryGenerator : ITweetQueryGenerator
    {
        private readonly IQueryParameterGenerator queryParameterGenerator;
        private readonly IUserQueryParameterGenerator userQueryParameterGenerator;

        public TweetQueryGenerator(IQueryParameterGenerator queryParameterGenerator, IUserQueryParameterGenerator userQueryParameterGenerator)
        {
            this.queryParameterGenerator = queryParameterGenerator;
            this.userQueryParameterGenerator = userQueryParameterGenerator;
        }

        // Get Tweet
        public string GetTweetQuery(IGetTweetParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Tweet_Get);

            query.AddParameterToQuery("id", parameters.Tweet?.Id.ToString() ?? parameters.Tweet?.IdStr);
            query.AddParameterToQuery("include_card_uri", parameters.IncludeCardUri);
            query.AddParameterToQuery("include_entities", parameters.IncludeEntities);
            query.AddParameterToQuery("include_ext_alt_text", parameters.IncludeExtAltText);
            query.AddParameterToQuery("include_my_retweet", parameters.IncludeMyRetweet);
            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetTweetsQuery(IGetTweetsParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Tweet_Lookup);

            var validTweetIdentifiers = parameters.Tweets.Where(x => GetTweetId(x) != null);
            var tweetIds = validTweetIdentifiers.Select(GetTweetId);

            query.AddParameterToQuery("id", string.Join(",", tweetIds));
            query.AddParameterToQuery("include_card_uri", parameters.IncludeCardUri);
            query.AddParameterToQuery("include_entities", parameters.IncludeEntities);
            query.AddParameterToQuery("include_ext_alt_text", parameters.IncludeExtAltText);
            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        // Publish Tweet
        public string GetPublishTweetQuery(IPublishTweetParameters parameters, TweetMode? requestTweetMode)
        {
            var text = parameters.Text;
            var useExtendedTweetMode = requestTweetMode == null || requestTweetMode == TweetMode.Extended;

            var quotedTweetUrl = GetQuotedTweetUrl(parameters);
            var attachmentUrl = parameters.QuotedTweetUrl;

            if (quotedTweetUrl != null)
            {
                // if there is a quoted tweet we need to pass the url in the text or attachment url
                // attachment_url is only available under tweetMode
                if (useExtendedTweetMode && attachmentUrl == null)
                {
                    attachmentUrl = quotedTweetUrl;
                }
                else
                {
                    text = text.TrimEnd() + " " + quotedTweetUrl;
                }
            }

            var query = new StringBuilder(Resources.Tweet_Publish);

            query.AddParameterToQuery("status", text);
            query.AddParameterToQuery("auto_populate_reply_metadata", parameters.AutoPopulateReplyMetadata);
            query.AddParameterToQuery("attachment_url", attachmentUrl);
            query.AddParameterToQuery("card_uri", parameters.CardUri);
            query.AddParameterToQuery("display_coordinates", parameters.DisplayExactCoordinates);

            if (parameters.ExcludeReplyUserIds != null)
            {
                query.AddParameterToQuery("exclude_reply_user_ids", string.Join(",", parameters.ExcludeReplyUserIds));
            }

            query.AddParameterToQuery("in_reply_to_status_id", GetTweetId(parameters.InReplyToTweet));
            query.AddParameterToQuery("lat", parameters.Coordinates?.Latitude.ToString(CultureInfo.InvariantCulture));
            query.AddParameterToQuery("long", parameters.Coordinates?.Longitude.ToString(CultureInfo.InvariantCulture));

            if (parameters.MediaIds.Count > 0)
            {
                var mediaIdsParameter = string.Join(",", parameters.MediaIds.Select(x => x.ToString(CultureInfo.InvariantCulture)));
                query.AddParameterToQuery("media_ids", mediaIdsParameter);
            }

            query.AddParameterToQuery("place_id", parameters.PlaceId);
            query.AddParameterToQuery("possibly_sensitive", parameters.PossiblySensitive);
            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);

            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        private string GetQuotedTweetUrl(IPublishTweetParameters parameters)
        {
            if (parameters.QuotedTweet?.CreatedBy?.ScreenName == null)
            {
                return null;
            }

            var quotedTweetId = GetTweetId(parameters.QuotedTweet);
            return $"https://twitter.com/{parameters.QuotedTweet.CreatedBy.ScreenName}/status/{quotedTweetId}";
        }

        public string GetDestroyTweetQuery(IDestroyTweetParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(string.Format(Resources.Tweet_Destroy, this.queryParameterGenerator.GenerateTweetIdentifier(parameters.Tweet)));

            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetFavoriteTweetsQuery(IGetUserFavoriteTweetsParameters parameters, TweetMode? requestTweetMode)
        {
            var userParameter = this.userQueryParameterGenerator.GenerateIdOrScreenNameParameter(parameters.User);
            var query = new StringBuilder(Resources.User_GetFavorites + userParameter);

            query.AddParameterToQuery("include_entities", parameters.IncludeEntities);
            this.queryParameterGenerator.AddMinMaxQueryParameters(query, parameters);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetRetweetsQuery(IGetRetweetsParameters parameters, TweetMode? requestTweetMode)
        {
            var tweetId = GetTweetId(parameters.Tweet);
            var query = new StringBuilder(string.Format(Resources.Tweet_Retweet_GetRetweets, tweetId));

            query.AddParameterToQuery("count", parameters.PageSize);
            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetPublishRetweetQuery(IPublishRetweetParameters parameters, TweetMode? requestTweetMode)
        {
            var tweetId = GetTweetId(parameters.Tweet);
            var query = new StringBuilder(string.Format(Resources.Tweet_Retweet_Publish, tweetId));

            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetDestroyRetweetQuery(IDestroyRetweetParameters parameters, TweetMode? requestTweetMode)
        {
            var tweetId = GetTweetId(parameters.Tweet);
            var query = new StringBuilder(string.Format(Resources.Tweet_DestroyRetweet, tweetId));

            query.AddParameterToQuery("trim_user", parameters.TrimUser);

            this.queryParameterGenerator.AppendTweetModeParameter(query, parameters.TweetMode ?? requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetRetweeterIdsQuery(IGetRetweeterIdsParameters parameters)
        {
            var query = new StringBuilder(Resources.Tweet_GetRetweeters);

            query.AddParameterToQuery("id", GetTweetId(parameters.Tweet));
            this.queryParameterGenerator.AppendCursorParameters(query, parameters);

            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        // Favorites
        public string GetCreateFavoriteTweetQuery(IFavoriteTweetParameters parameters)
        {
            var query = new StringBuilder(Resources.Tweet_Favorite_Create);

            query.AddParameterToQuery("id", GetTweetId(parameters.Tweet));
            query.AddParameterToQuery("include_entities", parameters.IncludeEntities);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetUnfavoriteTweetQuery(IUnfavoriteTweetParameters parameters)
        {
            var query = new StringBuilder(Resources.Tweet_Favorite_Destroy);

            query.AddParameterToQuery("id", GetTweetId(parameters.Tweet));
            query.AddParameterToQuery("include_entities", parameters.IncludeEntities);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        public string GetOEmbedTweetQuery(IGetOEmbedTweetParameters parameters)
        {
            var query = new StringBuilder(Resources.Tweet_GenerateOEmbed);

            query.AddParameterToQuery("id", GetTweetId(parameters.Tweet));
            query.AddParameterToQuery("maxwidth", parameters.MaxWidth);
            query.AddParameterToQuery("hide_media", parameters.HideMedia);
            query.AddParameterToQuery("hide_thread", parameters.HideThread);
            query.AddParameterToQuery("omit_script", parameters.OmitScript);
            query.AddParameterToQuery("align", this.queryParameterGenerator.GenerateOEmbedAlignmentParameter(parameters.Alignment));
            query.AddParameterToQuery("related", string.Join(",", parameters.RelatedUsernames ?? new string[0]));
            query.AddFormattedParameterToQuery(this.queryParameterGenerator.GenerateLanguageParameter(parameters.Language));
            query.AddParameterToQuery("theme", this.queryParameterGenerator.GenerateOEmbedThemeParameter(parameters.Theme));
            query.AddParameterToQuery("link_color", parameters.LinkColor);
            query.AddParameterToQuery("widget_type", parameters.WidgetType);
            query.AddParameterToQuery("dnt", parameters.EnablePersonalisationAndSuggestions);

            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        private string GetTweetId(ITweetIdentifier tweetIdentifier)
        {
            if (tweetIdentifier == null)
            {
                return null;
            }

            var tweetId = tweetIdentifier.IdStr;
            if (string.IsNullOrEmpty(tweetId))
            {
                tweetId = tweetIdentifier.Id.ToString(CultureInfo.InvariantCulture);
            }

            return tweetId;
        }

    }
}

File: Sharebook/Client/src/app/controllers/Tweet/TweetQueryGenerator.ts
import {ITweetQueryGenerator} from "../../core/Core/QueryGenerators/ITweetQueryGenerator";
import {ComputedTweetMode} from "../../core/Core/QueryGenerators/ComputedTweetMode";
import {Resources} from "../../properties/resources";
import {ITweetIdentifier} from "../../core/Public/Models/Interfaces/ITweetIdentifier";
import {TweetMode} from "../../core/Public/Settings/SharebookSettings";
import {IQueryParameterGenerator, IQueryParameterGeneratorToken} from "../Shared/QueryParameterGenerator";
import {
  IUserQueryParameterGenerator,
  IUserQueryParameterGeneratorToken
} from "../../core/Core/QueryGenerators/IUserQueryParameterGenerator";
import {IGetTweetParameters} from "../../core/Public/Parameters/TweetsClient/GetTweetParameters";
import {IGetTweetsParameters} from "../../core/Public/Parameters/TweetsClient/GetTweetsParameters";
import {IGetOEmbedTweetParameters} from "../../core/Public/Parameters/TweetsClient/GetOEmbedTweetParameters";
import {IUnfavoriteTweetParameters} from "../../core/Public/Parameters/TweetsClient/UnFavoriteTweetParameters";
import {IFavoriteTweetParameters} from "../../core/Public/Parameters/TweetsClient/FavoriteTweetParameters";
import {IGetRetweeterIdsParameters} from "../../core/Public/Parameters/TweetsClient/GetRetweeterIdsParameters";
import {IDestroyRetweetParameters} from "../../core/Public/Parameters/TweetsClient/DestroyRetweetParameters";
import {IPublishRetweetParameters} from "../../core/Public/Parameters/TweetsClient/PublishRetweetParameters";
import {IGetRetweetsParameters} from "../../core/Public/Parameters/TweetsClient/GetRetweetsParameters";
import {IGetUserFavoriteTweetsParameters} from "../../core/Public/Parameters/TweetsClient/GetFavoriteTweetsParameters";
import {IDestroyTweetParameters} from "../../core/Public/Parameters/TweetsClient/DestroyTweetParameters";
import {IPublishTweetParameters} from "../../core/Public/Parameters/TweetsClient/PublishTweetParameters";
import {Inject, Injectable} from "@angular/core";
import StringBuilder from "typescript-dotnet-commonjs/System/Text/StringBuilder";
import {StringBuilderExtensions} from "../../core/Core/Extensions/stringBuilder-extensions";
import {format} from "typescript-dotnet-commonjs/System/Text/Utility";

@Injectable({
  providedIn: 'root',
})
export class TweetQueryGenerator implements ITweetQueryGenerator {
  private readonly _queryParameterGenerator: IQueryParameterGenerator;
  private readonly _userQueryParameterGenerator: IUserQueryParameterGenerator;

  constructor(@Inject(IQueryParameterGeneratorToken) queryParameterGenerator: IQueryParameterGenerator,
              @Inject(IUserQueryParameterGeneratorToken) userQueryParameterGenerator: IUserQueryParameterGenerator) {
    this._queryParameterGenerator = queryParameterGenerator;
    this._userQueryParameterGenerator = userQueryParameterGenerator;
  }

  // Get Tweet
  public getTweetQuery(parameters: IGetTweetParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Tweet_Get);

    StringBuilderExtensions.addParameterToQuery(query, "id", parameters.tweet?.id.toString() ?? parameters.tweet?.idStr);
    StringBuilderExtensions.addParameterToQuery(query, "include_card_uri", parameters.includeCardUri);
    StringBuilderExtensions.addParameterToQuery(query, "include_entities", parameters.includeEntities);
    StringBuilderExtensions.addParameterToQuery(query, "include_ext_alt_text", parameters.includeExtAltText);
    StringBuilderExtensions.addParameterToQuery(query, "include_my_retweet", parameters.includeMyRetweet);
    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);
    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getTweetsQuery(parameters: IGetTweetsParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Tweet_Lookup);

    let validTweetIdentifiers = parameters.tweets.filter(x => this.getTweetId(x) != null);
    let tweetIds = validTweetIdentifiers.map(this.getTweetId);

    StringBuilderExtensions.addParameterToQuery(query, "id", tweetIds.join(', '));
    StringBuilderExtensions.addParameterToQuery(query, "include_card_uri", parameters.includeCardUri);
    StringBuilderExtensions.addParameterToQuery(query, "include_entities", parameters.includeEntities);
    StringBuilderExtensions.addParameterToQuery(query, "include_ext_alt_text", parameters.includeExtAltText);
    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);
    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  // Publish Tweet
  public getPublishTweetQuery(parameters: IPublishTweetParameters, tweetMode: ComputedTweetMode): string {
    let text = parameters.text;
    let useExtendedTweetMode = tweetMode === null || tweetMode.implicitOperator() === TweetMode.Extended;

    let quotedTweetUrl = this.getQuotedTweetUrl(parameters);
    let attachmentUrl = parameters.quotedTweetUrl;

    if (quotedTweetUrl != null) {
      // if there is a quoted tweet we need to pass the url in the text or attachment url
      // attachment_url is only available under tweetMode
      if (useExtendedTweetMode && attachmentUrl == null) {
        attachmentUrl = quotedTweetUrl;
      } else {
        text = text.trim() + " " + quotedTweetUrl;
      }
    }

    let query = new StringBuilder(Resources.Tweet_Publish);

    StringBuilderExtensions.addParameterToQuery(query, "status", text);
    StringBuilderExtensions.addParameterToQuery(query, "auto_populate_reply_metadata", parameters.autoPopulateReplyMetadata);
    StringBuilderExtensions.addParameterToQuery(query, "attachment_url", attachmentUrl);
    StringBuilderExtensions.addParameterToQuery(query, "card_uri", parameters.cardUri);
    StringBuilderExtensions.addParameterToQuery(query, "display_coordinates", parameters.displayExactCoordinates);

    if (parameters.excludeReplyUserIds != null) {
      StringBuilderExtensions.addParameterToQuery(query, "exclude_reply_user_ids", parameters.excludeReplyUserIds.join(',')); // string.Join(",", parameters.excludeReplyUserIds));
    }

    StringBuilderExtensions.addParameterToQuery(query, "in_reply_to_status_id", this.getTweetId(parameters.inReplyToTweet));
    StringBuilderExtensions.addParameterToQuery(query, "lat", parameters.coordinates?.latitude.toString(/*CultureInfo.InvariantCulture*/));
    StringBuilderExtensions.addParameterToQuery(query, "long", parameters.coordinates?.longitude.toString(/*CultureInfo.InvariantCulture*/));

    if (parameters.mediaIds.length > 0) {
      let mediaIdsParameter = parameters.mediaIds.map(x => x.toString(/*CultureInfo.InvariantCulture*/)).join(', ');
      StringBuilderExtensions.addParameterToQuery(query, "media_ids", mediaIdsParameter);
    }

    StringBuilderExtensions.addParameterToQuery(query, "place_id", parameters.placeId);
    StringBuilderExtensions.addParameterToQuery(query, "possibly_sensitive", parameters.possiblySensitive);
    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);
    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  private getQuotedTweetUrl(parameters: IPublishTweetParameters): string {
    if (parameters.quotedTweet?.createdBy?.screenName == null) {
      return null;
    }

    let quotedTweetId = this.getTweetId(parameters.quotedTweet);
    return `https://twitter.com/${parameters.quotedTweet.createdBy.screenName}/status/${quotedTweetId}`;
  }

  public getDestroyTweetQuery(parameters: IDestroyTweetParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(format(Resources.Tweet_Destroy, this._queryParameterGenerator.generateTweetIdentifier(parameters.tweet)));

    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.TrimUser);
    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getFavoriteTweetsQuery(parameters: IGetUserFavoriteTweetsParameters, tweetMode: ComputedTweetMode): string {
    let userParameter = this._userQueryParameterGenerator.generateIdOrScreenNameParameter(parameters.user);
    let query = new StringBuilder(Resources.User_GetFavorites + userParameter);

    StringBuilderExtensions.addParameterToQuery(query, "include_entities", parameters.includeEntities);
    this._queryParameterGenerator.addMinMaxQueryParameters(query, parameters);

    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getRetweetsQuery(parameters: IGetRetweetsParameters, tweetMode: ComputedTweetMode): string {
    let tweetId = this.getTweetId(parameters.tweet);
    let query = new StringBuilder(`https://api.twitter.com/1.1/statuses/retweets/${tweetId}.json`);

    StringBuilderExtensions.addParameterToQuery(query, "count", parameters.pageSize);
    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);

    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getPublishRetweetQuery(parameters: IPublishRetweetParameters, tweetMode: ComputedTweetMode): string {
    let tweetId = this.getTweetId(parameters.tweet);
    let query = new StringBuilder(format(Resources.Tweet_Retweet_Publish, tweetId));

    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);

    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getDestroyRetweetQuery(parameters: IDestroyRetweetParameters, tweetMode: ComputedTweetMode): string {
    let tweetId = this.getTweetId(parameters.tweet);
    let query = new StringBuilder(format(Resources.Tweet_DestroyRetweet, tweetId));

    StringBuilderExtensions.addParameterToQuery(query, "trim_user", parameters.trimUser);
    StringBuilderExtensions.addParameterToQuery(query, "tweet_mode", tweetMode);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getRetweeterIdsQuery(parameters: IGetRetweeterIdsParameters): string {
    let query = new StringBuilder(Resources.Tweet_GetRetweeters);

    StringBuilderExtensions.addParameterToQuery(query, "id", this.getTweetId(parameters.tweet));
    this._queryParameterGenerator.appendCursorParameters(query, parameters);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  // Favorites
  public getCreateFavoriteTweetQuery(parameters: IFavoriteTweetParameters): string {
    let query = new StringBuilder(Resources.Tweet_Favorite_Create);

    StringBuilderExtensions.addParameterToQuery(query, "id", this.getTweetId(parameters.tweet));
    StringBuilderExtensions.addParameterToQuery(query, "include_entities", parameters.includeEntities);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getUnfavoriteTweetQuery(parameters: IUnfavoriteTweetParameters): string {
    let query = new StringBuilder(Resources.Tweet_Favorite_Destroy);

    StringBuilderExtensions.addParameterToQuery(query, "id", this.getTweetId(parameters.tweet));
    StringBuilderExtensions.addParameterToQuery(query, "include_entities", parameters.includeEntities);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  public getOEmbedTweetQuery(parameters: IGetOEmbedTweetParameters): string {
    let query = new StringBuilder(Resources.Tweet_GenerateOEmbed);

    StringBuilderExtensions.addParameterToQuery(query, "id", this.getTweetId(parameters.tweet));
    StringBuilderExtensions.addParameterToQuery(query, "maxwidth", parameters.maxWidth);
    StringBuilderExtensions.addParameterToQuery(query, "hide_media", parameters.hideMedia);
    StringBuilderExtensions.addParameterToQuery(query, "hide_thread", parameters.hideThread);
    StringBuilderExtensions.addParameterToQuery(query, "omit_script", parameters.omitScript);
    StringBuilderExtensions.addParameterToQuery(query, "align", this._queryParameterGenerator.generateOEmbedAlignmentParameter(parameters.alignment));
    StringBuilderExtensions.addParameterToQuery(query, "related", parameters.relatedUsernames.join(', ') ?? new Array<string>(0).join(', ')); // string.Join(",", parameters.relatedUsernames ?? new string[0]));
    StringBuilderExtensions.addFormattedParameterToQuery(query, this._queryParameterGenerator.generateLanguageParameter(parameters.language));
    StringBuilderExtensions.addParameterToQuery(query, "theme", this._queryParameterGenerator.generateOEmbedThemeParameter(parameters.theme));
    StringBuilderExtensions.addParameterToQuery(query, "link_color", parameters.linkColor);
    StringBuilderExtensions.addParameterToQuery(query, "widget_type", parameters.widgetType);
    StringBuilderExtensions.addParameterToQuery(query, "dnt", parameters.enablePersonalisationAndSuggestions);

    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  private getTweetId(tweetIdentifier: ITweetIdentifier): string {
    if (tweetIdentifier == null) {
      return null;
    }

    let tweetId = tweetIdentifier.idStr;
    if (!tweetId) {
      tweetId = tweetIdentifier.id.toString(/*CultureInfo.InvariantCulture*/);
    }

    return tweetId;
  }
}
